src/main/java/com/redhat/rhjmc/containerjfr/net/web/WebServer.java

```diff
@@ -0,0 +1,384 @@
+        // blocking function, must be called from a blocking handler
+        try (recording) {
+            response.end(reportGenerator.generateReport(recording));
+
+            // ugly hack for "trimming" created clones of specified recording. JMC service creates a clone of running
+            // recordings before loading events to create the report, and these clones are erroneously left dangling.
+            TRIM_WORKER.submit(() -> {
+                try {
+                    service.getAvailableRecordings()
+                            .stream()
+                            .filter(r -> r.getName().equals(String.format("Clone of %s", recordingName)))
+                            .forEach(r -> {
+                                try {
+                                    service.close(r);
+                                } catch (FlightRecorderException fre) {
+                                    logger.debug(fre);
+                                }
+                            });
+                } catch (FlightRecorderException fre) {
+                    logger.debug(fre);
+                }
+            });
+        }
+    }
+
+    void handleClientUrlRequest(RoutingContext ctx) {
+        ctx.response().putHeader(HttpHeaders.CONTENT_TYPE, MIME_TYPE_JSON);
+        try {
+            endWithJsonKeyValue("clientUrl", String.format("ws://%s:%d/command", netConf.getWebServerHost(), netConf.getExternalWebServerPort()), ctx.response());
+        } catch (SocketException | UnknownHostException e) {
+            throw new HttpStatusException(500, e);
+        }
+    }
+
+    void handleGrafanaDatasourceUrlRequest(RoutingContext ctx) {
+        ctx.response().putHeader(HttpHeaders.CONTENT_TYPE, MIME_TYPE_JSON);
+        endWithJsonKeyValue("grafanaDatasourceUrl", env.getEnv(GRAFANA_DATASOURCE_ENV, ""), ctx.response());
+    }
+
+    void handleGrafanaDashboardUrlRequest(RoutingContext ctx) {
+        ctx.response().putHeader(HttpHeaders.CONTENT_TYPE, MIME_TYPE_JSON);
+        endWithJsonKeyValue("grafanaDashboardUrl", env.getEnv(GRAFANA_DASHBOARD_ENV, ""), ctx.response());
+    }
+
+    void handleRecordingDownloadRequest(String recordingName, RoutingContext ctx) {
+        try {
+            Optional<InputStream> recording = getRecordingInputStream(recordingName);
+            if (recording.isEmpty()) {
+                throw new HttpStatusException(404, String.format("%s not found", recordingName));
+            }
+
+            ctx.response().setChunked(true);
+            ctx.response().putHeader(HttpHeaders.CONTENT_TYPE, MIME_TYPE_OCTET_STREAM);
+            ctx.response().endHandler((e) -> downloadCounts.merge(recordingName, 1, Integer::sum));
+
+            if (System.getenv("USE_SYNC_WRITE_INPUT_STREAM") != null) {
+                writeInputStreamSync(recording.get(), ctx.response());
+                logger.info("useWriteInputStreamSync");
```

Comment:
Almost done with the review.

Since this check is based on an environment variable, which is constant after JVM startup, could you log the strategy used just once at WebServer creation or startup?

And, since both methods are blocking calls and really synchronous to the same degree (ie async only happens within the Response's write queue handling, but we block until all the data is copied there), I think this env var and the "sync" version of the method could use a rename.

Reply:
I agree. Will move to logging to the static initializer. Also, I'm wondering if you have any suggestions for renaming...

Reply:
Maybe something relating to "fixed size copy buffer"?

Reply:
If we're going to leave this as a configurable option then the name should be more indicative of the result than the implementation detail, so maybe the environment variable name should relate to "low memory pressure streaming", or some other keywords like that.

As for the names of the methods themselves, perhaps it could reflect that same environment variable naming. I'll try to think of some other options. "Sequential" came to mind, but I don't think that's really any more accurate than "synchronous" for this situation.

Reply:
Please see commit 23820bc.